src/dream.mli

```diff
@@ -1387,7 +1387,14 @@ val session_label : request -> string
 val session_expires_at : request -> float
 (** Time at which the session will expire. *)
 
+(** {1 Flash Messages} *)
 
+val flash_messages : ?lifetime:float -> middleware
+type level = Debug | Info | Success | Warning | Error
+
+val clear_flash : level -> request -> unit
+val put_flash : level -> string -> request -> unit
+val get_flash : level -> request -> string option
```

Comment:
Should this return a `string list`? Should it be possible to store multiple messages of the same level? Should `get_flash` return all messages of the same level or higher?

Reply:
> Should this return a string list? Should it be possible to store multiple messages of the same level?

Django allows you to store multiple messages of the same level; Phoenix appears to just let you store one (support for multiple messages was [added then removed](https://github.com/phoenixframework/phoenix/pull/304)). Looking in the [Phoenix issue](https://github.com/phoenixframework/phoenix/issues/301) linked earlier, one of the maintainers wrote:

> Going through my Ruby web apps, I almost always have a single flash message per key. The caveat of multiple is it makes the single case more verbose.

This, and the fact that multi-message support was added and then later removed makes me think the single message approach is better. But I'm happy to revise if you'd prefer the other approach.

> Should get_flash return all messages of the same level or higher?

I'm speculating, but I don't think most web developers treat flash messages' levels as totally ordered; they're more like labels for ensuring a particular message will be shown with relevant styling (green for info, red for error, etc.). It might have been better to choose some other word than "level", like "key" or "label".

Reply:
> This, and the fact that multi-message support was added and then later removed makes me think the single message approach is better. But I'm happy to revise if you'd prefer the other approach.

The comments since https://github.com/phoenixframework/phoenix/pull/304#issuecomment-421169891 are about adding multiple message support, and listing workarounds.

It seems worrying that flash messages can be lost &mdash; at least for me personally, it would give me anxiety. In more "technical" terms, it makes usage of flash message non-composable, since you cannot predict, purely locally, if a message you are adding will actually be displayed, despite there not being a `clear_flash`.

Out of the frameworks I checked/confirmed so far, Phoenix and Laravel don't support multiple messages. Django, Flask, Express, and Rails do, and expect you to iterate over the messages. It seems that the most used frameworks (Rails, Express, Django) use the multiple-messages style.

I personally don't see how code can be "correct" and *not* iterate over messages, given that there are already, at the very least, multiple levels, even if with only one message each. Wouldn't you just define a sub-template for doing this iteration, and base styles on each message's "level," which you get as an "output" of `get_flash` rather than pass as an "input?"

Is there a reason to ever hold on to some flash messages for display in response to a later request? Do you want to e.g. choose only the info messages, and not display errors?

I think the base flash functionality of Dream should be "correct" and worry-free, and we can build a single-message getter on top of it later, if it is needed.

Reply:
I also want to note that it is pretty easy for users to `List.filter` flash messages, or take the head, and it is easy to define the right helper based on a base functionality that doesn't ever lose messages. The converse is not true &mdash; it's not (easily) possible to define a multi-message behavior on top of a mutable, overwriting, single-message behavior, unless you wrap everything (not just the getter locally) so as to flatten lists into some kind of format inside strings, and parse them when retrieving.

Reply:
If you have experience with the Phoenix style of flash messages, could you explain the advantage of

```ocaml
val get_flash : category -> request -> message (option or list)
```

over 

```ocaml
val get_flash : request -> (category * message) list
```

What benefit does Phoenix/its users gain from choosing a single category, and passing it as argument?

Reply:
Reading this over in the last few days, I now think there are more advantages to the multi-message approach. I thought this point against the single-message approach was especially persuasive:

> it makes usage of flash message non-composable, since you cannot predict, purely locally, if a message you are adding will actually be displayed

I've revised to allow for multiple messages of the same category, and done away with fixed message categories.